### PR TITLE
Forms

### PR DESCRIPTION
**Pull request by IMGROOT2**
Forms Functionality.

Firebase rules have been deployed.

### CODE REVIEW

File: src/js/forms.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+const otherName = document.getElementById("name-other");
+const otherEmail = document.getElementById("email-other");
+const otherText = document.getElementById("text-other");
+const otherBtn = document.getElementById("btn-other");
+
+const toFeedback = document.getElementById("to-feedback");
+const toBug = document.getElementById("to-bug");
+const toTakedown = document.getElementById("to-takedown");
+const toOther = document.getElementById("to-other");
```
Comment (grimsteel) [inline src/js/forms.js:30]: collapse all of that into a forms object
kind of like this:
```js
const forms = {
  feedback: {...},
  bugs: {...},
  ...
};
```

File: src/js/forms.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        takedownBtn.addEventListener("click", async (e) => {
+            e.preventDefault();
+            takedownBtn.classList.add("is-loading");
+            await sendTakedown(user);
+        });
+        otherBtn.addEventListener("click", async (e) => {
+            e.preventDefault();
+            otherBtn.classList.add("is-loading");
+            await sendOther(user);
+        });
```
Comment (grimsteel) [inline src/js/forms.js:53]: along with the idea of a `forms` object with subobjects for each form, you could do something like this to reduce repetition:
```js
{
  feedback: {
    btn: ...,
    form: reference to native form element,
    function: sendFeedback
  }
}

Object.values(forms).forEach(form => form.form.addEventListener("submit", ...));

File: src/js/forms.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            location.href="/#login";
+        });
+        takedownBtn.addEventListener("click", async (e) => {
+            e.preventDefault();
+            location.href="/#login";
+        });
+        otherBtn.addEventListener("click", async (e) => {
+            e.preventDefault();
+            location.href="/#login";
+        });
```
Comment (grimsteel) [inline src/js/forms.js:87]: add a single submit event handler on the `main` element

for the "to" buttons, put them in an array and `forEach` it

File: src/js/forms.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        name: feedbackName.value,
+        email: feedbackEmail.value,
+        rating: feedbackRating.value,
+        text: feedbackText.value,
+        uid: auth.currentUser.uid
+    };
+    await addDoc(collection(db, "form_data", "types", "feedback"), feedback).then(() => {
+        feedbackBtn.classList.remove("is-loading");
+        feedbackBtn.disabled = true;
+        feedbackBtn.innerHTML = "Submitted!";
```
Comment (grimsteel) [inline src/js/forms.js:101]:
```suggestion
        feedbackBtn.innerText = "Submitted!";
```

File: src/forms.pug [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+                                input#email-other.input(type="email" name="email" placeholder="Your Email")
+                        .field
+                            .control
+                                label.label
+                                    | What's up?
+                                textarea#text-other.textarea(name="other" placeholder="Description")
+                        .field
+                            .control
+                                button#btn-other.button.is-primary
+                                    | Submit Other Request
```
Comment (grimsteel) [inline src/forms.pug:173]: some of these could be collapsed onto one line
```suggestion
                                button#btn-other.button.is-primary Submit Other Request
```

**Review by grimsteel — CHANGES_REQUESTED** [review]

File: src/forms.pug [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+                    a#to-feedback.button.is-primary.is-outlined(href="#feedback")
+                        | Submit Feedback
+            .tile.is-parent
+                .tile.is-child.box
+                        p.title
+                            | Bug Report
+                        p.subtitle
+                            | Vocabustudy is constantly in development, and a few bugs are sure to sneak in! If you found a bug, let us know! We'll fix it as soon as possible.
+                        p.subtitle.is-bold
+                            | We generally discourage use of this form, in favor of #[a(href="https://github.com/For-0/vocabustudy/issues") #[img(src="icons/brands/github.svg" style="height: 1em; vertical-align: middle; bottom: 2px; position: relative;")] GitHub Issues.] If you can, please use that instead.
```
Comment (grimsteel) [inline src/forms.pug:201]:
```suggestion
                            | We generally discourage use of this form, in favor of #[a.icon-text(href="https://github.com/For-0/vocabustudy/issues") #[span.icon.mr-2 #[img(src="icons/brands/github.svg")]] GitHub Issues.] If you can, please use that instead.
```

File: src/forms.pug [hunk 7]
```diff
@@ -19,17 +19,17 @@ block content
...
                                 label.label
                                     | Your Name
-                                input#name-feedback.input(type="text" name="name" placeholder="Your Name")
+                                input#name-feedback.input(type="text" name="name" placeholder="Your Name" required="required")
                         .field
                             .control
                                 label.label
                                     | Your Email - If you use your student email, you will not receive a follow up email.
-                                input#email-feedback.input(type="email" name="email" placeholder="Your Email")
+                                input#email-feedback.input(type="email" name="email" placeholder="Your Email" required="required")
```
Comment (grimsteel) [inline src/forms.pug:26]: This isn't XHTML :wink:
```suggestion
                                input#email-feedback.input(type="email" name="email" placeholder="Your Email" required)
```

File: src/js/forms.js [hunk 8]
```diff
@@ -29,65 +29,78 @@ const toBug = document.getElementById("to-bug");
...
-const { db, auth } = initialize(async user => {
+const feedbackForm = document.getElementById("feedback-form");
+const bugForm = document.getElementById("bug-form");
+const takedownForm = document.getElementById("takedown-form");
+const otherForm = document.getElementById("other-form");
+
+const {db, auth} = initialize(async user => {
     if (user) {
-        feedbackBtn.addEventListener("click", async (e) => {
+        feedbackForm.addEventListener('submit',  async (e) => {
```
Comment (grimsteel) [inline src/js/forms.js]: could you please modularize this? the repetition just decreases readability

**Review by grimsteel — CHANGES_REQUESTED** [review]

**Review by grimsteel — CHANGES_REQUESTED** [review]
Everything looks great, thanks! However the heroes are too big (esp. on chromebooks) so could you remove `.is-medium` from them? The content of the page under the heroes really matters more

Also add this:
```css
left: -1rem;
position: relative;
width: calc(100% + 2rem);
```
to all the heroes (maybe in the `extrahead` block?) to make them full width

**Review by grimsteel — APPROVED** [review]

**Comment by grimsteel** [comment]
contrary to what github says, I did _not_ approve the changes

this is why you always check where you are typing before pressing "enter"

File: src/js/forms.js (hunk 8 above)
Comment (IMGROOT2) [reply src/js/forms.js]: this is probably old because I did that already

File: firestore.rules [hunk 9]
```diff
@@ -61,5 +61,9 @@ service cloud.firestore {
     match /{path=**}/social/{uid} {
       allow list: if auth() && request.auth.uid == resource.data.uid;
     }
+    match /form_data/types/{test=**} {
+      allow get: if request.auth.token.admin == true;
+      allow create: if true;
+    }
```
Comment (grimsteel) [inline firestore.rules:67]: We should probably make this more secure (put the function at the top with the other functions):
```suggestion
    function validateFormResponse(additionalKeys) {
      let requiredFields = ["name", "email", "text", "uid"];
      return limitResourceKeys(requiredFields.concat(additionalKeys), []);
    }
    match /form_data/types/{form_type}/{submission_id} {
      allow get: if resource == null || isCreator();
      allow list, delete: if isCreator();
      allow create: if auth() && form_type in ["feedback", "bug", "takedown", "other"] && request.resource.data.keys().hasAll(["name", "email", "text", "uid])
        (form_type == "feedback" && validateFormResponse(["rating"])) ||
        (form_type == "feedback" && validateFormResponse(["url"])) ||
        (form_type == "feedback" && validateFormResponse(["url"])) ||
        (form_type == "feedback" && validateFormResponse([])));
    }
```

**Pull request merged**